### PR TITLE
Make s3 download by range

### PR DESCRIPTION
**Pull request by proggga**

### CODE REVIEW

**Comment by usernamedt** [comment]
Hi!

This repository has been merged into the wal-g main repo. Please re-submit it there.

Thanks!

File: s3/folder.go [hunk 1]
```diff
@@ -150,6 +155,129 @@ func (folder *Folder) PutObject(name string, content io.Reader) error {
 	return folder.uploader.upload(*folder.Bucket, folder.Path+name, content)
 }
 
+func (folder *Folder) CopyObject(srcPath string, dstPath string) error {
```
Comment (x4m) [inline s3/folder.go:158]: This seem to be carbon copy of a code in another place.

File: s3/folder.go [hunk 2]
```diff
@@ -150,6 +155,129 @@ func (folder *Folder) PutObject(name string, content io.Reader) error {
...
+
+	tracelog.DebugLogger.Printf("Init s3reader id %s path %s", reader.id, objectPath)
+	return reader
+}
+
+func getId(objectPath string, counter int) string {
+	hash := fnv.New32a()
+	_, err := hash.Write([]byte(objectPath))
+	if err != nil {
+		tracelog.ErrorLogger.Println("Fatal, can't write buffer to hash")
```
Comment (x4m) [inline s3/folder.go:275]: Do we have a more idiomatic ways to fatal\panic in tracelogger?

File: s3/folder.go [hunk 3]
```diff
@@ -150,6 +155,129 @@ func (folder *Folder) PutObject(name string, content io.Reader) error {
...
+				return errors.Wrap(err, "s3Reader Too much reconnecting retries")
+			}
+			time.Sleep(time.Duration(sleepMultiplier) * backoffSleepDuration)
+			sleepMultiplier *= 2
+			continue
+		}
+		if reader.lastBody != nil {
+			err = reader.lastBody.Close()
+			if err != nil {
+				return errors.Wrap(err, "s3Reader We have problems with closing prev connection, smth strange")
```
Comment (x4m) [inline s3/folder.go:248]: ", smth strange" part is extraneous :)

File: s3/folder.go [hunk 4]
```diff
@@ -150,6 +155,129 @@ func (folder *Folder) PutObject(name string, content io.Reader) error {
...
+
+	for  {
+		object, err := reader.getObjectRange(reader.storageCursor, 0)
+		if err != nil {
+			failed += 1
+			tracelog.DebugLogger.Printf("s3Reader [%s] reconnect failed: %s", reader.id, err)
+			if failed >= reader.maxRetries {
+				return errors.Wrap(err, "s3Reader Too much reconnecting retries")
+			}
+			time.Sleep(time.Duration(sleepMultiplier) * backoffSleepDuration)
```
Comment (x4m) [inline s3/folder.go:241]: How many retries do we allow? Do we have an upper limit of retry sleep? We may need a jitter. Maybe use build-in Retrier from S3 SDK?

**Pull request closed without merging**